### ISSUE TITLE
Pure null check in binary operator doesn't affect 'else' branch

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#47291]
**Version Used**: e7bad77

**Steps to Reproduce**: [SharpLab link](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+BiAdgVwBtCJhC4ACOXU8gWACgABAZgqYCYKBhCgb0YUh7NkwAsFALIAKJgEYADBQhJ2iisACU/QcL0BLAGYVpECvqgUCxCgDJbG85euFNuvUIEMPwgL7uPOEIoOAC9Lx8fCAA6ABUAewBlGAR9XABzaU0AbgoAejyKAHcIBFw09LDI4DiklIqs3IKKKAALeKIUYtLcVWB8GAoUeLgoXAByGCqhf29ImITk1IzG/MKSsorpjVqlhpy17s2MgNnfIA===)

```cs
#nullable enable
public class C {
    public void M(string a, string b) {
        if (a is null && b is null)
        {
        }
        else
        {
            a.ToString(); // warning
            b.ToString(); // should warn, but doesn't
        }
        a.ToString(); // warning
        b.ToString(); // warning
    }
}
```

I don't know why 'a' is considered maybe-null in the else branch here but not 'b'. One key detail to note is that the declared types are non-nullable.

**Comment by Youssef1313** [comment]
Interesting bug. Any idea where I should start debugging if I want to give it a try?

**Comment by RikkiGibson** [comment]
Check [`NullableWalker.LearnFromNullTest`](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs#L3801) (I think) and [`NullableWalker.VisitBinaryOperatorChildren`](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs#L3384). I would debug and add a watch for `this`, step through, and see how it is that we make it into the `else` block in a state where `b` is not-null.

**Comment by Youssef1313** [comment]
Will give it a try

**Comment by Youssef1313** [comment]
I put a break-point in the first line of `VisitBinaryOperatorChildren`, but it wasn't hit at all.

**Comment by RikkiGibson** [comment]
Weird, maybe there's actually a different code path for `&&`. Given that I would break on [`ReportPossibleNullReceiverIfNeeded`](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs#L9084) and check the call stack, or if even that fails break on [`DiagnosticBag.Add`](https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/Diagnostic/DiagnosticBag.cs#L116) 😉
